### ISSUE TITLE
Graph model fails to replicate on multiple devices. 

### ISSUE THREAD

**Issue by nasimrahaman** [nasimrahaman/inferno#3]
As of [this commit](https://github.com/nasimrahaman/inferno/tree/f87058b033797807e2decfbac12f25496a726e62), the problem can be reproduced as follows: 
```python
import torch
from torch.autograd import Variable
import torch.nn as nn
from torch.nn.parallel.data_parallel import data_parallel
from inferno.extensions.containers.graph import Graph

input_shape = [8, 1, 3, 128, 128]
model = Graph()\
    .add_input_node('input')\
    .add_node('conv0', nn.Conv3d(1, 10, 3, padding=1), previous='input')\
    .add_node('conv1', nn.Conv3d(10, 1, 3, padding=1), previous='conv0')\
    .add_output_node('output', previous='conv1')

model.cuda()
input = Variable(torch.rand(*input_shape).cuda())
output = data_parallel(model, input, device_ids=[0, 1, 2, 3])
```

This raises: 
```
RuntimeError: tensors are on different GPUs
```

Could this be due to [this `add_module`](https://github.com/nasimrahaman/inferno/blob/f87058b033797807e2decfbac12f25496a726e62/inferno/extensions/containers/graph.py#L205)?

**Comment by nasimrahaman** [comment]
Tested with Pytorch V0.2, still does not work. :disappointed:

**Comment by nasimrahaman** [comment]
The bug was due to model._graph containing copies of the module, fixed in 3e3613f1cb4e16d0bbc27b05ae4f62d0580757f9.

**Comment by nasimrahaman** [comment]
The bug is a more subtle threading issue. Due to [this line in `torch.nn.parallel.replicate`](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/replicate.py#L29), the [`Graph._graph`](https://github.com/nasimrahaman/inferno/blob/master/inferno/extensions/containers/graph.py#L70) object is not replicated. Replicas from [multiple threads](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/parallel_apply.py#L50) (i.e. multiple devices) therefore end up reading from and writing to the same networkx graph object, which is bad. 

The solution would be to annotate the payload with thread-id. Using a threading lock (to concurrently access `Graph._graph`) isn't completely trivial because they're not pickle-able, but that can be solved by overloading `torch.nn.Module`'s `__getitem__` and `__setitem__` methods.

**Comment by nasimrahaman** [comment]
Fixed with #35.
